Simplify trust store and trust policy. (#149)
notaryproject/notaryproject: trust-store-trust-policy-specification.md

# Trust Store and Trust Policy Specification

Notary v2 currently supports X.509 based PKI and identities, and uses a trust store and trust policy to determine if a signed artifact is considered authentic.

The document consists of the following sections:

> - **[Trust Store](#trust-store)**: Contains a set of trusted identities through which trust is derived for the rest of the system. For X509 PKI, the trust store contains a set of root certificates.

Reviewer: This is a question.  Is an intermediary Certificate allowed in this collection and thus not always in the true sense a root certificate.

Reviewer: Notice below you changed the verbiage to say typically.

Author: Yes, intermediate certs are allowed. Updated.
```suggestion
- **[Trust Store](#trust-store)**: Contains a set of trusted identities through which trust is derived for the rest of the system. For X.509 PKI, the trust store typically contains a set of root certificates.
```

> - **[Trust Store](#trust-store)**: Contains a set of trusted identities through which trust is derived for the rest of the system. For X.509 PKI, the trust store typically contains a set of root certificates.

Reviewer 2: nit, "typically" are there times when it wouldn't be root certs?

- **[Trust Policy](#trust-policy)**: A policy language which indicates which identities are trusted to produce artifacts. Both trust store and trust policy need to be configured by users/administrators before artifact signature can be evaluated.
- **[Signature Verification](#signature-verification)**: Describes how signatures are evaluated using the policy, to determine if a signed artifact is authentic. This section is meant for implementors of Notary v2 standards.

Other types of identities and trust models may be supported in future, which may introduce other constructs/policy elements to support signature evaluation.

All examples use the actors defined in Notary v2 [scenario](https://github.com/notaryproject/notaryproject/blob/main/scenarios.md#scenario-0-build-publish-consume-enforce-policy-deploy)

> 

Reviewer 2: A suggestion to put a header that says: ## Scenario

- Wabbit Networks company builds, signs and distributes their `net-monitor` software though public registries.

> - ACME Rockets consumes the `net-monitor` software from a public registry importing the  artifacts and reference artifacts (signatures, SBoMs) into their private registry. The private registry also contains additional artifacts that ACME Rockets themselves sign.

Reviewer 2: nit the <double space> artifacts

## Trust Store

Contains a set of trusted identities through which trust is derived for the rest of the system. For X509 PKI, the trust store typically contains root certificates.

- The Notary v2 trust store consists of multiple named collections of certificates, called named stores.
- Following certificate formats are supported - Files with extension .pem, .crt and .cer, the files are expected to contain certificate(s) in DER (binary) format or PEM format (base-64 encoded DER).

> - The trust store is a directory location, under which each sub directory is considered a named store, that contains zero or more certificates. The name of this sub directory is used to reference the specfic store in trust policy.

Reviewer: <sp> specfic => specific

Author: ```suggestion
- The trust store is a directory location, under which each sub directory is considered a named store, that contains zero or more certificates. The name of this sub directory is used to reference the specific store in trust policy.
```

> - Symlinks are not supported for the named store directories or certificate files. Implementation MUST validate that the named store directory or certificate files are not symlinks, and fail it it does not meet this condition.

Reviewer 2: ...fail it it 
should be:
...fail **if** it

```jsonc
{
    "version": "1.0",
    "trustPolicies": [
        {
            # Policy for set of artifacts signed by Wabbit Networks
            # that are pulled from ACME Rockets repository
            "name": "wabbit-networks-images",
            "registryScopes": [
                "registry.acme-rockets.io/software/net-monitor"
                "registry.acme-rockets.io/software/net-logger" ],
            "signatureVerification" : "strict",
            "trustedIdentities": {
                "trustStore": "ca:acme-rockets",
                "identities": [ 
                  "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io, OU=Security Tools"
                ]
            }
        },
        {
            # Exception policy for a single unsigned artifact pulled from
            # Wabbit Networks repository
            "name": "unsigned-image",
            "registryScopes": [ "registry.wabbit-networks.io/software/unsigned/net-utils" ],
            "signatureVerification": "skip",
        },
        {
            # Policy for with custom verification policy

>             "name": "use-expired-image",

Reviewer 2: Should we require * to be added to support all?
Or, should we assume a default scenario where a an entire registry is trusted, unless scoped?

I realize that someone may want to limit docker hub content to library, however most secured environments use private registries, where the registry is the trust boundary, not necessarily a sub-namespace.

Author: The `registryScopes` with value `*` makes it explicit to user that any registry is trusted, it's better to be explicit in security related policies, unlike other policies where we may favor conciseness.  We used the same for `identities` where it's now a required property but you can specify a wildcard `*`.

```
"name": "myPolicy",
            "registryScopes": [ "*" ],
            "signatureVerification" : "strict",
            "trustedIdentities": {
                "trustStore": "ca:acme-rockets",
                "identities": [ "*"]
            }
```

- The Notary v2 trust store consists of multiple named collections of certificates, called named stores.
- The trust store is a directory location, under which each sub directory is considered a named store, that contains zero or more certificates. The name of this sub directory is used to reference the specfic store in trust policy.
- Certificates in a trust store are typically root certificates. Intermediate certificates can be used, but not recommended as intermediate certificates can be rotated more often and can break signature verification.

Notary v2 uses following directory structure to represent the trust store. The example shows named stores `acme-rockets` and `wabbit-networks`, which are subseqently references in the trust policy. Without this reference, presence of a named store and certificates in it does not confer trust automatically to the named store. The trust store is configured ahead of verification time, by an out of band mechanism that is beyond the scope of this document. Different entities and organizations have their own processes and policies to configure and distribute trust stores.

```text

> ~/.notation/trust_store

Reviewer 3: The notation CLI currently follows the [XDG](https://wiki.archlinux.org/title/XDG_Base_Directory) directories. Should the root directory be `$XDG_CONFIG_HOME/notation/trust_store`?

Author: Should it be `$XDG_CONFIG_HOME/.notation/trust_store` instead?

> ~/.notation/trust_store

Reviewer 3: It seems that it is more common in Linux to have hyphens than underscores. For instances, `trust-store` and `ca-roots`.

    /x509
        /ca_roots
            /acme-rockets

>                 cert1.pem

Reviewer 3: Is the extension restricted to `.pem`? Is `.crt` accepted?

Author: Addressed, what about `.cer`?

                cert2.pem
                  /sub-dir       # ignored
                    cert-3.pem   # ignored
            /wabbit-networks
                cert3.pem
        /tsa_roots
            /publicly-trusted-tsa
                tsa-cert1.pem
```

The Trust store currently supports two kinds of identities, additional identities may be supported in future :

> - **Certificates**: The `x509/ca` trust store contains named stores that contain Certificate Authority (CA) root and intermediate certificates.

Reviewer 3: `x509/ca` is inconsistent with `/x509/ca_roots`.

Author: Fixed in latest version.

> - **Certificates**: The `x509/ca` trust store contains named stores that contain Certificate Authority (CA) root and intermediate certificates.

Reviewer 3: Intermediate certificates should not be in the `roots` folders since they are not root CA certs.

- **Timestamping Certificates**: The `x509/tsa_roots` trust store contains named stores with Time Stamping Authority (TSA) root and intermediate certificates. **NOTE** TSA based timestamping will not be available in Notation RC1.

Any additional sub directories under names stores and certificates in it are ignored.

## Trust Policy

Describes how signatures are evaluated using the policy, to determine if a signed artifact is authentic. Users who consume and execute the signed artifact from a registry need a mechanism to specify how the artifacts should be evaluated for trust, this is where a trust policy is used.
Trust policy allows users to control the artifact's integrity, expiry, and revocation aspect of signature evaluation.

### Trust Policy Schema

The trust policy is as JSON document, example shown below:

```jsonc
{
    "version": "1.0",
    "trustPolicies": [
        {
            "name": "policy-for-wabbit-networks-images",
            "registryScopes": [
                "registry.acme-rockets.io/software/net-monitor"
                "registry.acme-rockets.io/software/net-logger" ],
            "signatureVerification" : "strict",

>             "trustStore": "ca/wabbit-networks",

Reviewer 3: I'm confused here since `ca` does not match `ca_roots`. Should we make everything consistent?

Author: Fixed the typo in latest version.

            "trustedIdentity": [
                "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io"
            ],
        },
        {
            "name": "policy-for-unsigned-image",

>             "registryScopes": [ "registry.wabbit-networks.io/software/unsigned/productA" ],

Reviewer 3: nit: repository name does not allow capital letters.

|skip      |Use to exclude verification for unsigned images|Skip|Skip|Skip|Skip|Skip|

**Integrity** : Guarantees that the artifact wasn't altered after it was signed. All signature verifications levels always enforce integrity. Invalid signatures are rejected.
**Authenticity** : Guarantees that the artifact was signed by an identity trusted by the verifier. Its definition does not include revocation, which is when a trusted identity is subsequently untrusted because of a compromise.
**Trusted timestamp** : Guarantees that the signature was generated when the certificate was valid. It also allows a verifier to determine if a signature be treated as valid when a certificate is revoked, if the certificate was revoked after the signature was generated. In the absence of a trusted timestamp, signatures are considered invalid after certificate expires, and all signatures are considered revoked when a certificate is revoked.
**Expiry** : This is an optional feature that gurantees that an artifact is considered fresh, if the signing identity indicates an optional artifact expiry time.

> **Revocation check** : Guarantees that the signing identity is still trusted at signature verification time. Events such as key or system compromise can make a signing identity that was previously trusted, to be subsequrntly untrusted. This guarantee typically requires a verification time call to an external system, which may not be consistently reliable. The `permissive` verification level only audits for revocation check and does not enforce it. If a particular revocation mechnanism provides is reliable, use `strict` verification level instead.

Reviewer 3: We probably need empty lines between those lines for better rendering.

  - **`trustedIdentities`**(*object*): This REQUIRED property specifies a set of identities that the user trusts. For X.509 PKI, trusted identities is specified with a combination of trust store and trust anchors.
    - **`trustStore`**(*string*): This REQUIRED property specifies a named trust store. Uses the format `{trust-store-type}:{named-store}`. Currently supported values for `trust-store-type` are `ca` and `tsa`. **NOTE**: When support for publicly trusted TSA is available, `tsa:publicly-trusted-tsa` is the default value, and implied without explictly specifying it. If a custom TSA is used the format `ca:acme-rockets,tsa:acme-tsa` is supported to specify it.
    - **`identities`**(*array of strings*): This REQUIRED property specifies a list of elements/attributes of the signing certificate's subject. For more information, see [trusted identities constraints](#trusted-identities-constraints) section. A value `*` is supported if user trusts any identity (signing certificate) issued by the CA(s) in `trustStore`.

#### Signature Verification details

> Notary v2 defines the following signature verification levels, to provide different levels of enforcement for different scenarios. Signature verrification is a multi step process that checks for the following - integrity (artifact is unaltered, signature is not currupted), authenticity (the signature is really from the identity that claims to have signed it), trusted timestamping (the signature was generated when the key/certificate were unexpired), expiry (an optional check if the artifact specifies an expiry date), revocation check (is the signing identity still trusted at present time). Based on the signature verification level, each of these checks is `enforced` or `logged`. `Enforced` causes signature verification to fail, and `logged` causes failures to be logged and the next check is evaluated. NOTE: Implementations may change the ordering of these checks, but all checks MUST to be performed for signature verification process to be considered complete.

Reviewer: <sp> currupted => corrupted

> Notary v2 defines the following signature verification levels, to provide different levels of enforcement for different scenarios. Signature verrification is a multi step process that checks for the following - integrity (artifact is unaltered, signature is not currupted), authenticity (the signature is really from the identity that claims to have signed it), trusted timestamping (the signature was generated when the key/certificate were unexpired), expiry (an optional check if the artifact specifies an expiry date), revocation check (is the signing identity still trusted at present time). Based on the signature verification level, each of these checks is `enforced` or `logged`. `Enforced` causes signature verification to fail, and `logged` causes failures to be logged and the next check is evaluated. NOTE: Implementations may change the ordering of these checks, but all checks MUST to be performed for signature verification process to be considered complete.

Reviewer: Two comments.  Normally expiry check is cheap and done first, and the MUST in the last sentence is odd.  Normally at the first failure you stop, but do you want to continue the rest of the checks to perform a full audit?  That though means potentially megacycles of CPU.

Author: That's good feedback, the last sentence isn't clear enough. Reworded and addressed typos.

```suggestion
Notary v2 defines the following signature verification levels, to provide different levels of enforcement for different scenarios. Signature verification is a multi step process performs the following validation checks - integrity (artifact is unaltered, signature is not corrupted), authenticity (the signature is really from the identity that claims to have signed it), trusted timestamping (the signature was generated when the key/certificate were unexpired), expiry (an optional check if the artifact specifies an expiry time), revocation check (is the signing identity still trusted at the present time). Based on the signature verification level, each of these validation checks is `enforced` or `logged`. If a validation check is `enforced`, a failure is treated as critical failure, and causes the overall signature verification to fail.  If a validation check is `logged`, a failure causes details to be logged, and the next validation check is evaluated till all checks succeed or a critical failure is encountered. NOTE: Implementations may change the ordering of these checks based on efficiency, but all validation checks MUST to be performed till the first critical failure is encountered, or all checks succeed, for signature verification process to be considered complete.
```

The following table shows the resultant behavior `enforced` (verification fails), or `logged` for each of the checks, based on signature verification level.

|Signature Verification Level|Recommended Usage|Integrity|Authenticity|Trusted timestamp|Expiry|Revocation check|
|----------------------------|-----------------|---------|------------|-----------------|------|----------------|
|strict    |Use at development, build and deploy time|enforced|enforced|enforced|enforced|enforced|
|permissive|Use at deploy time or runtime|enforced|enforced|logged|logged|logged|
|audit     |Use when adopting signed images, without breaking existing workflows|enforced|logged|logged|logged|logged|
|skip      |Use to exclude verification for unsigned images|skipped|skipped|skipped|skipped|skipped|

**Integrity** : Guarantees that the artifact wasn't altered after it was signed, or the signature isn't corrupted. All signature verification levels always enforce integrity.

**Authenticity** : Guarantees that the artifact was signed by an identity trusted by the verifier. Its definition does not include revocation, which is when a trusted identity is subsequently untrusted because of a compromise.

> **Trusted timestamp** : Guarantees that the signature was generated when the certificate was valid. It also allows a verifier to determine if a signature be treated as valid when a certificate is revoked, if the certificate was revoked after the signature was generated. In the absence of a trusted timestamp, signatures are considered invalid after certificate expires, and all signatures are considered revoked when a certificate is revoked. **NOTE**: `notation` RC1 will generate trusted timestamp using a TSA when signature is generated, but will not support verification of TSA countersignatures.

Reviewer: <grammar> TSA when signature => TSA when the signature

Author: ```suggestion
**Trusted timestamp** : Guarantees that the signature was generated when the certificate was valid. It also allows a verifier to determine if a signature be treated as valid when a certificate is revoked, if the certificate was revoked after the signature was generated. In the absence of a trusted timestamp, signatures are considered invalid after certificate expires, and all signatures are considered revoked when a certificate is revoked. **NOTE**: `notation` RC1 will generate trusted timestamp using a TSA when the signature is generated, but will not support verification of TSA countersignatures.
```

> **Expiry** : This is an optional feature that guarantees that artifact is within “best by use” date indicated in the signature. Notary v2 allows users to include an optional expiry time when they generate a signature. The expiry time is not set by default and requires explicit configuration by users at the time of signature generation. The artifact is considered expired after the expiry time has elapsed, users performing verification can either configure their trust policies to fail the verification or even accept the signature with expiry date in the past using policy. This is an advanced feature that allows implementing controls for user defined semantics like deprecation for older artifacts, or block older artifacts in a production environment. Users should only include an expiry time in the signed artifact after considering the behavior they expect for consumers of the artifact after it expires. Users can choose to consume an artifact even after the expiry time based on their specific needs.

Reviewer: expiry time has elapsed is unclear.  Do you mean when the current time is greater than or equal to the expiry time or just greater?

Author: Updated wording to be precise - current time is greater than or equal to expiry time
```suggestion
**Expiry** : This is an optional feature that guarantees that artifact is within “best by use” date indicated in the signature. Notary v2 allows users to include an optional expiry time when they generate a signature. The expiry time is not set by default and requires explicit configuration by users at the time of signature generation. The artifact is considered expired when the current time is greater than or equal to expiry time, users performing verification can either configure their trust policies to fail the verification or even accept the artifact with expiry date in the past using policy. This is an advanced feature that allows implementing controls for user defined semantics like deprecation for older artifacts, or block older artifacts in a production environment. Users should only include an expiry time in the signed artifact after considering the behavior they expect for consumers of the artifact after it expires. Users can choose to consume an artifact even after the expiry time based on their specific needs.
```

  1. *Global*: If there exists a trust policy with global scope then use that policy for signature evaluation.
     Otherwise, fail the signature verification.

### Trusted Identities Constraints

A distinguished name (usually just shortened to "DN") uniquely identifies an entry and in the case of the certificate's subject, DN uniquely identifies the requestor/holder of the certificate.

> The DN is comprised of zero or more comma-separated components called relative distinguished names, or RDNs.

Reviewer: in the line below do you mean to have '"' at the end of org1?

Author: Fixed in latest revision.

- Trust anchor MUST support a full and partial list of all the attribute types present in [subject DN](https://www.rfc-editor.org/rfc/rfc5280.html#section-4.1.2.6) of x509 certificate.
- If the subject DN of the signing certificate is used in the trust anchor, then it MUST meet the following requirements:
  - The value of `trustedIdentities` MUST begin with `x509.subject:` followed by comma-separated one or more RDNs.
    For example, `x509.subject: C=${country}, ST=${state}, L=${locallity}, O={organization}, OU=${organization-unit}, CN=${common-name}`.
  - Trust anchor MUST contain country (CN), state Or province (ST), and organization (O) RDNs.
    All other RDNs are optional.
    The minimal possible trust anchor is `x509.subject: C=${country}, ST=${state}, O={organization}`,

>   - Trust anchor MUST NOT have overlapping values.

Reviewer: What is the expected error behavior when this happens.  Does the first one win or does functionality stop, or does the attempt to update this policy fail?

Author: Good point, I'll clarify that signature verification fails with "invalid policy".
